common/credsregistry/provider.go

```diff
@@ -0,0 +1,146 @@
+// Therefore, explicitly state via arguments.
+func NewJSONProviderCredentials(
+	filePath string,
+	withAccessToken bool,
+) (*ProviderCredentials, error) {
+	return createProviderCreds(getProviderName(filePath), withAccessToken, filePath)
+}
+
+// NewENVProviderCredentials reads ENV variables associated with a provider.
+func NewENVProviderCredentials(
+	providerName string,
+	withAccessToken bool,
+) (*ProviderCredentials, error) {
+	return createProviderCreds(providerName, withAccessToken, "")
+}
+
+func createProviderCreds(
+	providerName string, withAccessToken bool, filePath string,
+) (*ProviderCredentials, error) {
+	// load provider from catalog to imply fields in JSON or ENV vars
+	catalog, err := providers.ReadCatalog()
+	if err != nil {
+		return nil, errors.Join(err, ErrProviderNotFound)
+	}
+
+	info, ok := catalog[providerName]
+	if !ok {
+		return nil, ErrProviderNotFound
+	}
+
+	fields, err := getFields(info, withAccessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	registry := NewCredentialsRegistry()
+	readers := handy.Lists[Reader]{}
+
+	// add readers for every field
+	for kind, fieldList := range fields {
+		for _, field := range fieldList {
+			reader := selectReader(field, filePath, providerName)
+			readers.Add(kind, reader)
+
+			if err = registry.AddReader(reader); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	r := &ProviderCredentials{
+		Registry:       registry,
+		ProviderValues: make(map[string]string),
+	}
+
+	return r, r.loadValues(readers)
+}
+
+func selectReader(field Field, filePath string, providerName string) Reader { // nolint:ireturn
+	if len(filePath) != 0 {
+		return field.GetJSONReader(filePath)
+	}
+
+	return field.GetENVReader(providerName)
+}
+
+func (r ProviderCredentials) loadValues(readers handy.Lists[Reader]) error {
+	// validate JSON file or ENV has all Required variables
+	missingKeys := make([]string, 0)
+
+	for _, reader := range readers["required"] {
+		key, err := reader.Key()
+		if err != nil {
+			return err
+		}
+
+		if value, err := r.Registry.GetString(key); err != nil {
+			missingKeys = append(missingKeys, key)
+		} else {
+			r.ProviderValues[key] = value
+		}
+	}
+
+	if len(missingKeys) != 0 {
+		return fmt.Errorf("%w: %s", ErrKeyNotFound, strings.Join(missingKeys, ","))
+	}
+
+	// Optional values should be saved too
+	for _, reader := range readers["optional"] {
+		key, err := reader.Key()
+		if err != nil {
+			return err
+		}
+
+		if value, err := r.Registry.GetString(key); err == nil {
+			r.ProviderValues[key] = value
+		}
+	}
+
+	return nil
+}
+
+func (r ProviderCredentials) Get(field Field) string {
+	return r.ProviderValues[field.Name]
+}
+
+func getProviderName(filePath string) string {
+	registry := NewCredentialsRegistry()
+
+	reader := Fields.Provider.GetJSONReader(filePath)
+
+	err := registry.AddReader(reader)
+	if err != nil {
+		return ""
+	}
+
+	name, _ := registry.GetString(Fields.Provider.Name)
```

Comment:
This registry can only read provider name from a file. Then lookup the catalog and imply the rest of fields.